pkg/volume/glusterfs/glusterfs.go

```diff
@@ -0,0 +1,233 @@
+}
+
+type glusterfsPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &glusterfsPlugin{}
+
+const (
+	glusterfsPluginName = "kubernetes.io/glusterfs"
+)
+
+func (plugin *glusterfsPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *glusterfsPlugin) Name() string {
+	return glusterfsPluginName
+}
+
+func (plugin *glusterfsPlugin) CanSupport(spec *api.Volume) bool {
+	if spec.VolumeSource.Glusterfs != nil {
+		return true
+	}
+	return false
+}
+
+func (plugin *glusterfsPlugin) GetAccessModes() []api.AccessModeType {
+	return []api.AccessModeType{
+		api.ReadWriteOnce,
+		api.ReadOnlyMany,
+		api.ReadWriteMany,
+	}
+}
+
+func (plugin *glusterfsPlugin) NewBuilder(spec *api.Volume, podRef *api.ObjectReference) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, podRef, mount.New(), exec.New())
+}
+
+func (plugin *glusterfsPlugin) newBuilderInternal(spec *api.Volume, podRef *api.ObjectReference, mounter mount.Interface, exe exec.Interface) (volume.Builder, error) {
+	return &glusterfs{
+		volName: spec.Name,
+		hosts:   spec.VolumeSource.Glusterfs.Hosts,
+		path:    spec.VolumeSource.Glusterfs.Path,
+		option:  spec.VolumeSource.Glusterfs.MountOpt,
+		helper:  spec.VolumeSource.Glusterfs.Helper,
+		mounter: mounter,
+		exe:     exe,
+		podRef:  podRef,
+		plugin:  plugin,
+	}, nil
+}
+
+func (plugin *glusterfsPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, mount.New())
+}
+
+func (plugin *glusterfsPlugin) newCleanerInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &glusterfs{
+		volName: volName,
+		mounter: mounter,
+		podRef:  &api.ObjectReference{UID: podUID},
+		plugin:  plugin,
+	}, nil
+}
+
+// Glusterfs volumes represent a bare host file or directory mount of an Glusterfs export.
+type glusterfs struct {
+	volName string
+	podRef  *api.ObjectReference
+	hosts   []string
+	path    string
+	option  string
+	helper  string
+	mounter mount.Interface
+	exe     exec.Interface
+	plugin  *glusterfsPlugin
+}
+
+// SetUp attaches the disk and bind mounts to the volume path.
+func (glusterfsVolume *glusterfs) SetUp() error {
+	return glusterfsVolume.SetUpAt(glusterfsVolume.GetPath())
+}
+
+func (glusterfsVolume *glusterfs) SetUpAt(dir string) error {
+	mountpoint, err := mount.IsMountPoint(dir)
+	glog.V(4).Infof("Glusterfs: mount set up: %s %v %v", dir, mountpoint, err)
+	if err != nil && !os.IsNotExist(err) {
+		return err
+	}
+	if mountpoint {
+		return nil
+	}
+	path := glusterfsVolume.path
+	os.MkdirAll(dir, 0750)
+	err = glusterfsVolume.execMount(glusterfsVolume.hosts, path, dir, glusterfsVolume.option, glusterfsVolume.helper)
+	if err != nil {
+		mountpoint, mntErr := mount.IsMountPoint(dir)
+		if mntErr != nil {
+			glog.Errorf("Glusterfs: IsMountpoint check failed: %v", mntErr)
+			return err
+		}
+		if mountpoint {
```

Comment:
it seems other volume plugins are doing the same. we'll have another project to refactor common code.